adapters/AppLovin/AppLovinAdapter/GADMAdapterAppLovinNative.m

```diff
@@ -0,0 +1,159 @@
+}
+
+- (void)getNativeAdWithAdTypes:(NSArray *)adTypes options:(NSArray *)options {
+  if (!([adTypes containsObject:kGADAdLoaderAdTypeUnifiedNative]
+        || [adTypes containsObject:kGADAdLoaderAdTypeNativeAppInstall])) {
+    NSError *error = [NSError errorWithDomain: GADMAdapterAppLovinConstant.errorDomain
+                                         code: kGADErrorInvalidRequest userInfo: nil];
+    [self.connector adapter:self didFailAd:error];
+    return;
+  }
+
+  self.adTypes = adTypes;
+
+  [[ALSdk shared].nativeAdService loadNativeAdGroupOfCount:1 andNotify:self];
+}
+
+- (void)stopBeingDelegate {
+  self.connector = nil;
+}
+
+#pragma mark - AppLovin Native Load Delegate Methods
+
+- (void)nativeAdService:(nonnull ALNativeAdService *)service
+didFailToLoadAdsWithError
+                       :(NSInteger)code {
+  [self notifyFailureWithErrorCode:[GADMAdapterAppLovinUtils toAdMobErrorCode:(int)code]];
+}
+
+- (void)nativeAdService:(nonnull ALNativeAdService *)service didLoadAds:(nonnull NSArray *)ads {
+  if (ads.count < 1) {
+    [self notifyFailureWithErrorCode:kGADErrorMediationNoFill];
+    return;
+  }
+
+  [service precacheResourcesForNativeAd:[ads firstObject] andNotify:self];
+}
+
+#pragma mark - AppLovin Native Ad Precache Delegate Methods
+
+- (void)nativeAdService:(nonnull ALNativeAdService *)service
+didFailToPrecacheImagesForAd
+                       :(nonnull ALNativeAd *)ad
+              withError:(NSInteger)errorCode {
+  [self notifyFailureWithErrorCode:[GADMAdapterAppLovinUtils toAdMobErrorCode:(int)errorCode]];
+}
+
+- (void)nativeAdService:(nonnull ALNativeAdService *)service
+didFailToPrecacheVideoForAd
+                       :(nonnull ALNativeAd *)ad
+              withError:(NSInteger)errorCode {
+  [self notifyFailureWithErrorCode:[GADMAdapterAppLovinUtils toAdMobErrorCode:(int)errorCode]];
+}
+
+- (void)nativeAdService:(nonnull ALNativeAdService *)service
+ didPrecacheImagesForAd:(nonnull ALNativeAd *)ad {
+  if ([self.adTypes containsObject:kGADAdLoaderAdTypeUnifiedNative]
+      && [GADMAdapterAppLovinUtils containsRequiredUnifiesNativeAssets:ad]) {
+    GADMAppLovinMediatedNativeUnifiedAd *unifiedNativeAd =
+        [[GADMAppLovinMediatedNativeUnifiedAd alloc] initWithNativeAd:ad];
+    if (unifiedNativeAd) {
+      [self.connector adapter:self didReceiveMediatedUnifiedNativeAd:unifiedNativeAd];
+      return;
+    }
+  } else if ([self.adTypes containsObject:kGADAdLoaderAdTypeNativeAppInstall]
+             && [GADMAdapterAppLovinUtils containsRequiredAppInstallNativeAssets:ad]) {
+    GADMAppLovinMediatedNativeAppInstallAd *appInstallNativeAd =
+        [[GADMAppLovinMediatedNativeAppInstallAd alloc] initWithNativeAd:ad];
+    if (appInstallNativeAd) {
+      [self.connector adapter:self didReceiveMediatedNativeAd:appInstallNativeAd];
+      return;
+    }
+  }
+
+  [self notifyFailureWithErrorCode:kGADErrorNoFill];
```

Comment:
Is it possible that `self.adTypes` contains multiple values?

Reply:
Also, what bothers me is the utility methods to check whether the native ad contains valid assets is in the utiltiy class here, but on Android, it is not. We try to keep both platforms as similar to each other as possible, for easier debugging and readability. Also these methods are only used in one place so I'd probably keep it here.